### PR TITLE
fix: validate status codes

### LINKED ISSUE #13: Validate status codes
Per https://github.com/ipfs/kubo/pull/8890#discussion_r971339756

> Please move this validation logic to Parse func in [https://github.com/ipfs/go-ipfs-redirects-file/blob/main/redirects.go](vscode-file://vscode-app/Applications/Visual%20Studio%20Code.app/Contents/Resources/app/out/vs/code/electron-sandbox/workbench/workbench.html).
> Since we have our own library, all the validation and safety checks should happen there, if possible.

### PR DESCRIPTION
**Pull request by justindotpub**
Closes #13

### CODE REVIEW

File: redirects_test.go [hunk 1]
```diff
@@ -58,12 +58,22 @@ func TestRule_IsRewrite(t *testing.T) {
...
 }
 
-func TestParse(t *testing.T) {
+func Test_Parse(t *testing.T) {
 	t.Run("with illegal force", func(t *testing.T) {
 		_, err := Parse(strings.NewReader(`
 		/home / 301!
 		`))
 
-		assert.Error(t, err, "forced redirects should return an error")
```
Comment (justindotpub) [inline redirects_test.go:67]: This test wasn't working.

File: redirects_test.go [hunk 2]
```diff
@@ -58,12 +58,22 @@ func TestRule_IsRewrite(t *testing.T) {
...
+		assert.Contains(t, err.Error(), "forced redirects")
+	})
+
+	t.Run("with illegal code", func(t *testing.T) {
+		_, err := Parse(strings.NewReader(`
+		/home / 42
+		`))
+
+		assert.Error(t, err)
+		assert.Contains(t, err.Error(), "status code 42 is not supported")
```
Comment (justindotpub) [inline redirects_test.go:77]: One token bad status.  I avoided returning a URL or a full list of status codes that are supported. 🤷

File: redirects.go [hunk 3]
```diff
@@ -158,5 +158,23 @@ func parseStatus(s string) (code int, err error) {
...
+
+	if !isValidStatusCode(code) {
+		return code, fmt.Errorf("status code %d is not supported", code)
+	}
+
 	return code, err
 }
+
+func isValidStatusCode(status int) bool {
+	validStatuses := []int{200, 301, 302, 303, 307, 308, 404, 410, 451}
```
Comment (lidel) [inline redirects.go]: This should be a const at the top of the file, no readon to allocate this every time

File: redirects.go [hunk 4]
```diff
@@ -158,5 +158,23 @@ func parseStatus(s string) (code int, err error) {
 	}
 
 	code, err = strconv.Atoi(s)
+	if err != nil {
+		return code, err
```
Comment (lidel) [inline redirects.go]: nit: no value if an error occurred, removes surface for bugs when err is ignored
```suggestion
		return nil, err
```

File: redirects.go [hunk 5]
```diff
@@ -158,5 +158,23 @@ func parseStatus(s string) (code int, err error) {
...
+	}
+
+	if !isValidStatusCode(code) {
+		return code, fmt.Errorf("status code %d is not supported", code)
```
Comment (lidel) [inline redirects.go]:
```suggestion
		return nil, fmt.Errorf("status code %d is not supported", code)
```

File: redirects.go [hunk 6]
```diff
@@ -158,5 +158,23 @@ func parseStatus(s string) (code int, err error) {
...
+	}
+
 	return code, err
```
Comment (lidel) [inline redirects.go]:
```suggestion
	return code, nil
```

**Review by lidel — CHANGES_REQUESTED** [review]
lgtm, but mind addressibg below @justincjohnson ?

File: redirects.go (hunk 4 above)
Comment (justindotpub) [reply redirects.go]: I can't return nil because it's an int.  I'll return 0 though.

File: redirects.go (hunk 3 above)
Comment (justindotpub) [reply redirects.go]: Done.  It had to be a var since slices can't be const's in Go.

**Comment by justindotpub** [comment]
@lidel All feedback has been addressed and pushed.

**Comment by justindotpub** [comment]
FYI, I've rebased the other PRs with these changes as well.

**Review by lidel — APPROVED** [review]
Thanks, simplified it even further.

**Pull request merged**